File: tools/extract_video_frames.py
from pathlib import Path
import cv2

input_dir = Path("public/assets/Video Animation")
output_dir = Path("tmp/video-frames")
output_dir.mkdir(parents=True, exist_ok=True)

rows = []
for video_path in sorted(input_dir.glob("*.mp4")):
    cap = cv2.VideoCapture(str(video_path))
    if not cap.isOpened():
        rows.append((video_path.name, "ERROR", "", "", ""))
        continue

    fps = cap.get(cv2.CAP_PROP_FPS) or 0
    frame_count = cap.get(cv2.CAP_PROP_FRAME_COUNT) or 0
    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH) or 0)
    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT) or 0)
    duration = frame_count / fps if fps else 0
    safe_name = "".join(ch if ch.isalnum() else "-" for ch in video_path.stem).strip("-")

    for second in (1, 5, 9):
        target = min(second, max(duration - 0.25, 0))
        cap.set(cv2.CAP_PROP_POS_MSEC, target * 1000)
        ok, frame = cap.read()
        if ok:
            output_path = output_dir / f"{safe_name}-{second}s.png"
            cv2.imwrite(str(output_path), frame)

    cap.release()
    rows.append((video_path.name, f"{duration:.2f}s", f"{fps:.2f}", f"{width}x{height}", f"{int(frame_count)} frames"))

print("Name\tDuration\tFPS\tResolution\tFrames")
for row in rows:
    print("\t".join(row))

File: tools/extract-video-frames.mjs
import { chromium } from 'playwright-core';
import fs from 'node:fs/promises';
import path from 'node:path';
import { pathToFileURL } from 'node:url';

const inputDir = path.resolve('public/assets/Video Animation');
const outputDir = path.resolve('tmp/video-frames');
await fs.mkdir(outputDir, { recursive: true });

const candidates = [
  process.env.PLAYWRIGHT_CHROMIUM_EXECUTABLE,
  'C:/Program Files/Google/Chrome/Application/chrome.exe',
  'C:/Program Files (x86)/Google/Chrome/Application/chrome.exe',
  'C:/Program Files/Microsoft/Edge/Application/msedge.exe',
  'C:/Program Files (x86)/Microsoft/Edge/Application/msedge.exe',
].filter(Boolean);

let executablePath = null;
for (const candidate of candidates) {
  try {
    await fs.access(candidate);
    executablePath = candidate;
    break;
  } catch {
    // try next
  }
}

if (!executablePath) {
  throw new Error('No Chromium/Edge executable found.');
}

const files = (await fs.readdir(inputDir)).filter((file) => file.toLowerCase().endsWith('.mp4'));
const browser = await chromium.launch({ executablePath });
const page = await browser.newPage({ viewport: { width: 960, height: 540 }, deviceScaleFactor: 1 });

for (const file of files) {
  const fileUrl = pathToFileURL(path.join(inputDir, file)).href;
  await page.setContent(`
    <!doctype html>
    <html>
      <body style="margin:0;background:#111;display:grid;place-items:center;width:960px;height:540px;overflow:hidden">
        <video id="video" src="${fileUrl}" muted playsinline style="width:960px;height:540px;object-fit:contain;background:#111"></video>
      </body>
    </html>
  `);

  await page.waitForFunction(() => {
    const video = document.querySelector('video');
    return video && Number.isFinite(video.duration) && video.duration > 0;
  });

  for (const second of [1, 5, 9]) {
    await page.evaluate((time) => {
      const video = document.querySelector('video');
      video.currentTime = Math.min(time, Math.max(video.duration - 0.25, 0));
    }, second);
    await page.waitForFunction(() => {
      const video = document.querySelector('video');
      return video.readyState >= 2;
    });
    await page.waitForTimeout(250);

    const safeName = file.replace(/[^a-z0-9]+/gi, '-').replace(/^-+|-+$/g, '');
    await page.screenshot({ path: path.join(outputDir, `${safeName}-${second}s.png`) });
  }
}

await browser.close();
console.log(`Extracted ${files.length * 3} frames to ${outputDir}`);
